muxer.go

```diff
@@ -0,0 +1,411 @@
+package astits
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"github.com/asticode/go-astikit"
+	"io"
+)
+
+const (
+	StartPID           uint16 = 0x0100
+	PMTStartPID        uint16 = 0x1000
+	ProgramNumberStart uint16 = 1
+)
+
+var (
+	MuxerErrorPIDNotFound      = errors.New("PID not found")
+	MuxerErrorPIDAlreadyExists = errors.New("PID already exists")
+	MuxerErrorPCRPIDInvalid    = errors.New("PCR PID invalid")
+)
+
+type Muxer struct {
+	ctx        context.Context
+	w          io.Writer
+	bitsWriter *astikit.BitsWriter
+
+	packetSize             int
+	tablesRetransmitPeriod int // period in PES packets
+
+	pm         programMap // pid -> programNumber
+	pmt        PMTData
+	nextPID    uint16
+	patVersion wrappingCounter
+	pmtVersion wrappingCounter
+
+	patBytes bytes.Buffer
+	pmtBytes bytes.Buffer
+
+	buf       bytes.Buffer
+	bufWriter *astikit.BitsWriter
+
+	esContexts              map[uint16]*esContext
+	tablesRetransmitCounter int
+}
+
+type esContext struct {
+	es *PMTElementaryStream
+	cc wrappingCounter
+}
+
+func newEsContext(es *PMTElementaryStream) *esContext {
+	return &esContext{
+		es: es,
+		cc: newWrappingCounter(0b1111), // CC is 4 bits
+	}
+}
+
+func MuxerOptionTablesRetransmitPeriod(newPeriod int) func(*Muxer) {
+	return func(m *Muxer) {
+		m.tablesRetransmitPeriod = newPeriod
+	}
+}
+
+func NewMuxer(ctx context.Context, w io.Writer, opts ...func(*Muxer)) *Muxer {
+	m := &Muxer{
+		ctx: ctx,
+		w:   w,
+
+		packetSize:             MpegTsPacketSize, // no 192-byte packet support yet
+		tablesRetransmitPeriod: 40,
+
+		pm: newProgramMap(),
+		pmt: PMTData{
+			ElementaryStreams: []*PMTElementaryStream{},
+			ProgramNumber:     ProgramNumberStart,
+		},
+
+		// table version is 5-bit field
+		patVersion: newWrappingCounter(0b11111),
+		pmtVersion: newWrappingCounter(0b11111),
+
+		esContexts: map[uint16]*esContext{},
+	}
+
+	m.bufWriter = astikit.NewBitsWriter(astikit.BitsWriterOptions{Writer: &m.buf})
+	m.bitsWriter = astikit.NewBitsWriter(astikit.BitsWriterOptions{Writer: m.w})
+
+	// TODO multiple programs support
+	m.pm.set(PMTStartPID, ProgramNumberStart)
+
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	// to output tables at the very start
+	m.tablesRetransmitCounter = m.tablesRetransmitPeriod
+
+	return m
+}
+
+// if es.ElementaryPID is zero, it will be generated automatically
+func (m *Muxer) AddElementaryStream(es PMTElementaryStream, isPCRPid bool) error {
```

Comment:
I wouldn't add `isPCRPid bool` in this function. Instead I would create a `func (m *Muxer) setPCRPID(pid uint16)` specifically for that

Reply:
Fair enough

Reply:
Done